src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -135,6 +163,50 @@ static int GetUnescapedOffset(ReadOnlySpan<char> fullName, int startIndex)
             }
         }
 
+        internal static string Unescape(string input)
+        {
+            int indexOfEscapeChar = input.IndexOf(EscapeCharacter);
+            if (indexOfEscapeChar < 0)
+            {
+                // Nothing to escape, just return the original value.
+                return input;
+            }
+
+            ValueStringBuilder builder = new(stackalloc char[128]);
+            builder.EnsureCapacity(input.Length);
+
+            UnescapeToBuilder(input.AsSpan(), ref builder);
+
+            string result = builder.ToString();
+            builder.Dispose();
+            return result;
+
+            static void UnescapeToBuilder(ReadOnlySpan<char> input, ref ValueStringBuilder builder)
+            {
+                while (!input.IsEmpty)
+                {
+                    int indexOfEscapeChar = input.IndexOf(EscapeCharacter);
+                    if (indexOfEscapeChar < 0)
+                    {
+                        builder.Append(input);
+                        break;
+                    }
+                    builder.Append(input.Slice(0, indexOfEscapeChar));
+                    int indexOfNextChar = indexOfEscapeChar + 1;
+                    if (indexOfNextChar < input.Length && input[indexOfNextChar] is char c && NeedsEscaping(c))
```

Comment:
You should not need to call NeedsEscaping here. Take a look how the existing Unescape helper in TypeNameHelpers.cs is implemented.

Reply:
Given that this is very rare case, it is better to keep the implementation as simple as possible - the existing Unescape helper in TypeNameHelpers.cs that processes one character at a time looks simpler.

Reply:
Copied implementation from `TypeNameHelpers.cs`, removed it and updated usages.